rotkehlchen/chain/evm/decoding/metamask/decoder.py

```diff
@@ -0,0 +1,177 @@
+from rotkehlchen.chain.evm.decoding.structures import (
+    DEFAULT_DECODING_OUTPUT,
+    DecoderContext,
+    DecodingOutput,
+)
+from rotkehlchen.chain.evm.decoding.types import CounterpartyDetails
+from rotkehlchen.chain.evm.decoding.utils import maybe_reshuffle_events
+from rotkehlchen.chain.evm.types import string_to_evm_address
+from rotkehlchen.history.events.structures.types import HistoryEventSubType, HistoryEventType
+from rotkehlchen.types import ChecksumEvmAddress
+from rotkehlchen.utils.misc import hex_or_bytes_to_address, hex_or_bytes_to_int
+
+from .constants import (
+    CPT_METAMASK,
+    METAMASK_FEE_TOPIC,
+    METAMASK_ICON,
+    METAMASK_LABEL,
+    SWAP_SIGNATURE,
+)
+
+if TYPE_CHECKING:
+    from rotkehlchen.chain.evm.decoding.base import BaseDecoderTools
+    from rotkehlchen.chain.evm.node_inquirer import EvmNodeInquirer
+    from rotkehlchen.user_messages import MessagesAggregator
+
+
+class MetamaskCommonDecoder(DecoderInterface):
+
+    def __init__(
+            self,
+            evm_inquirer: 'EvmNodeInquirer',
+            base_tools: 'BaseDecoderTools',
+            msg_aggregator: 'MessagesAggregator',
+            router_address: str,
+            gnosis_safe_proxy: bytes,
+            native_currency: Asset,
+    ) -> None:
+        super().__init__(evm_inquirer, base_tools, msg_aggregator)
+        self.router_address = string_to_evm_address(router_address)
+        self.gnosis_safe_proxy = gnosis_safe_proxy
+        self.native_currency = native_currency
+
+    def _decode_swap(self, context: DecoderContext) -> DecodingOutput:
+        """
+        This function is used to decode the swap done by metamask.
+        By reading the contract code the spending address and the receiving address
+        seem to be the same, so it doesn't check for that.
+        """
+
+        if context.tx_log.topics[0] != SWAP_SIGNATURE:
+            return DEFAULT_DECODING_OUTPUT
+
+        sender = hex_or_bytes_to_address(context.tx_log.topics[2])
+        if not self.base.is_tracked(sender):
+            return DEFAULT_DECODING_OUTPUT
+
+        out_event = in_event = None
+        for event in context.decoded_events:
+            asset_symbol = event.asset.resolve_to_asset_with_symbol().symbol
+            if (
+                event.event_type == HistoryEventType.TRADE and
+                event.event_subtype == HistoryEventSubType.SPEND or
+                event.event_type == HistoryEventType.SPEND and
+                event.event_subtype == HistoryEventSubType.NONE and
+                event.location_label == sender
+            ):
+                # find the send event
+                event.counterparty = CPT_METAMASK
+                event.event_type = HistoryEventType.TRADE
+                event.event_subtype = HistoryEventSubType.SPEND
+                event.notes = f'Swap {event.balance.amount} {asset_symbol} in {CPT_METAMASK} from {event.location_label}'  # noqa: E501
```

Comment:
no need to write to/from. This is already visible in the event

```suggestion
                event.notes = f'Swap {event.balance.amount} {asset_symbol} in {CPT_METAMASK}'
```